single_kernel_mongo/managers/mongos_operator.py

```diff
@@ -0,0 +1,318 @@
+        if name != RelationNames.MONGOS_PROXY:
+            return False
+        return True
+
+    def share_connection_info(self):
+        """Shares the connection information of clients."""
+        if not self.state.db_initialised:
+            return
+        if not self.charm.unit.is_leader():
+            return
+        try:
+            self._share_configuration()
+        except PyMongoError as e:
+            raise DeferrableError(f"updating app relation data because of {e}")
+        except ApiError as e:  # Raised for k8s
+            if e.status.code == 404:
+                raise DeferrableError(
+                    "updating app relation data since service not found for more or one units"
+                )
+            raise
+
+    def _share_configuration(self):
+        """Actually shares the configuration according to the substrate."""
+        match self.substrate:
+            case Substrates.VM:
+                if not self.state.mongos_config.password or not self.state.mongos_config.username:
+                    return
+                for relation in self.state.client_relations:
+                    self.mongo_manager.update_app_relation_data_for_config(
+                        relation, self.state.mongos_config
+                    )
+            case Substrates.K8S:
+                for relation in self.state.client_relations:
+                    self.mongo_manager.update_app_relation_data(relation)
+
+    def share_credentials(self, relation: Relation):
+        """Shares credentials to the client."""
+        data_interface = DatabaseProviderData(self.model, relation.name)
+        if not self.charm.unit.is_leader():
+            return
+        new_database_name = data_interface.fetch_relation_field(relation.id, "database")
+        new_extra_user_roles: set[str] = set(
+            json.loads(
+                data_interface.fetch_relation_field(
+                    relation.id,
+                    "extra-user-roles",
+                )
+                or "[]"
+            )
+        )
+        external_connectivity = json.loads(
+            data_interface.fetch_relation_field(relation.id, "external-node-conectivity") or "false"
+        )
+
+        if new_database_name and new_database_name != self.state.app_peer_data.database:
+            self.state.app_peer_data.database = new_database_name
+            if self.state.mongos_cluster_relation:
+                self.state.cluster.database = new_database_name
+
+        if new_extra_user_roles != self.state.app_peer_data.extra_user_roles:
+            self.state.app_peer_data.extra_user_roles = new_extra_user_roles
+            if self.state.mongos_cluster_relation:
+                self.state.cluster.extra_user_roles = new_extra_user_roles
+
+        self.state.app_peer_data.external_connectivity = external_connectivity
+
+        if external_connectivity:
+            self.charm.unit.open_port("tcp", MongoPorts.MONGOS_PORT)
```

Comment:
hmm this seems like a side effect to the function which is supposed to only share credentials, is there a better place to put this?